### PR TITLE
Implement synchronized interactive linked axes for PyPlot and Plotly.

### PR DESCRIPTION
**Pull request by apalugniok**
#1371 

Can't really make an image for this so please try it out. I have a weird error where if the code added is at the start of the loop over the subplots it doesn't work but it's fine at the end? Ever since I updated matplotlib PyPlot has been a bit weird so I'm not sure if this error is just me.

### CODE REVIEW

File: src/backends/pyplot.jl [hunk 1]
```diff
@@ -1139,6 +1139,11 @@ function _before_layout_calcs(plt::Plot{PyPlotBackend})
         # this sets the bg color inside the grid
         ax[set_facecolor_sym](py_color(sp[:background_color_inside]))
 
+        # link axes
+        if plt[:link] != :none
+            ax[:get_shared_x_axes]()[:join](ax, sp[:xaxis].sps[1].o)
+            ax[:get_shared_y_axes]()[:join](ax, sp[:yaxis].sps[1].o)
```
Comment (mkborregaard) [inline src/backends/pyplot.jl]: why do you only check if `link != :none`? Other possibilities are `:x`, `:y`, `:both` and `:all`

Comment (apalugniok) [reply src/backends/pyplot.jl]: I don't want to link the axes when `link` isn't directly specified as the other possibilities you mentioned. Linking the axes to themselves caused issues with polar plots and I think heatmaps... I still need to check this out properly.

File: src/backends/plotly.jl [hunk 2]
```diff
@@ -448,9 +454,15 @@ const _plotly_markers = KW(
     :hline      => "line-ew",
 )
 
-function plotly_subplot_index(sp::Subplot)
-    spidx = sp[:subplot_index]
-    spidx == 1 ? "" : spidx
+# find indicies of axes to which the supblot links to
+function plotly_link_indicies(plt::Plot, sp::Subplot)
```
Comment (apalugniok) [inline src/backends/plotly.jl]: Is there a way to access the `Plot` from a `Subplot`, I couldn't find anything 😟

Comment (apalugniok) [inline src/backends/plotly.jl]: Is there a way to access the `Plot` from a `Subplot`, I couldn't find anything 😟

File: src/backends/plotly.jl [hunk 3]
```diff
@@ -448,9 +454,15 @@ const _plotly_markers = KW(
...
+    if plt[:link] in (:x, :y, :both)
```
Comment (apalugniok) [inline src/backends/plotly.jl]: Plotly can't support `link = :all` with synchronized interactivity should we add a warning about this here?

File: src/backends/plotly.jl (hunk 2 above)
Comment (daschw) [reply src/backends/plotly.jl]: Does `sp.plt` work in this context?

Comment (apalugniok) [reply src/backends/plotly.jl]: Yeah that's what I was looking for, not sure how I missed that possibility 😆

Comment (apalugniok) [reply src/backends/plotly.jl]: Although now that  think about it it seems clearer to have the `Plot` in the function calls if it's attributes are going to be used. So I think it's fine as it is now 😄

**Comment by cstjean** [comment]
This feature is awesome, is the PR missing anything significant? I was sold on the [demo](https://github.com/JuliaPlots/Plots.jl/issues/1371#issuecomment-365637156)!

**Comment by apalugniok** [comment]
I simply haven't found the time to test it properly before merging (mainly how it interacts with the new `ticks = :native`). I'll rebase so that others can easily see if it works. I can't think of any glaring errors but they always sneak up on you 😆

**Comment by mkborregaard** [comment]
> is the PR missing anything significant?

I guess it's only really missing someone taking it for a spin round the corners of Plots

**Comment by cstjean** [comment]
Once it's rebased, I'll gladly give it a test drive.

**Comment by apalugniok** [comment]
Should be ready for a test drive now :D

**Comment by cstjean** [comment]
I'm not sure how to do a gif, but on OSX + plotly, this code is broken. Panning the top plot moves the bottom plot's axis (but not the top one), and vice versa :stuck_out_tongue: Also note the line segment sticking out of its plotting area, at the bottom left of the image.

image

**Comment by apalugniok** [comment]
I can't reproduce this (only tried in REPL) D: I will have time to properly checkout this issue next week!

**Comment by cstjean** [comment]
That's strange, I can reproduce at the REPL. Julia 0.6, OSX, Plots commit da4111579

```julia
julia> using Plots

julia> plotlyjs()
Plots.PlotlyJSBackend()

julia> plot(plot([1,2,3]),
            plot([3,4,5]), link=:x, layout=(2,1))
```

**Comment by apalugniok** [comment]
So I've had no luck reproducing any of this D: Does anybody else get the same behavior?

**Comment by daschw** [comment]
I can reproduce on Windows10, but I remember that I could not on archlinux ... will check again today in the evening.

**Comment by apalugniok** [comment]
Could you also share the output html that plotly produces? I can then check it against what I get.

**Comment by daschw** [comment]
Apparently it works for me on Windows with `plotly()`
```html
<!DOCTYPE html>
<html>
    <head>
        <title>Plots.jl</title>
        <script src="file://C:\Users\Daniel\.julia\v0.6\Plots\src\backends\..\..\deps\plotly-latest.min.js"></script>
    </head>
    <body>
            <div id="971acd9f-c7fd-43c3-98a0-b92a9bf3f4c9" style="width:600px;height:400px;"></div>
    <script>
    PLOT = document.getElementById('971acd9f-c7fd-43c3-98a0-b92a9bf3f4c9');
    Plotly.plot(PLOT, [{"showlegend":true,"mode":"lines","xaxis":"x1","colorbar":{"title":""},"line":{"color":"rgba(0, 154, 250, 1.000)","shape":"linear","dash":"solid","width":1},"y":[1,2,3],"type":"scatter","name":"y1","yaxis":"y1","x":[1,2,3]},{"showlegend":true,"mode":"lines","xaxis":"x1","colorbar":{"title":""},"line":{"color":"rgba(0, 154, 250, 1.000)","shape":"linear","dash":"solid","width":1},"y":[3,4,5],"type":"scatter","name":"y1","yaxis":"y2","x":[1,2,3]}], {"paper_bgcolor":"rgba(255, 255, 255, 1.000)","height":400,"yaxis2":{"showticklabels":true,"gridwidth":0.5,"tickvals":[3.0,3.5,4.0,4.5,5.0],"visible":true,"ticks":"inside","range":[3.0,5.0],"domain":[0.03762029746281716,0.4901574803149606],"tickmode":"array","linecolor":"rgba(0, 0, 0, 1.000)","showgrid":true,"title":"","mirror":false,"tickangle":0,"showline":true,"gridcolor":"rgba(0, 0, 0, 0.100)","titlefont":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":15},"tickcolor":"rgb(0, 0, 0)","ticktext":["3.0","3.5","4.0","4.5","5.0"],"zeroline":false,"type":"-","tickfont":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":11},"zerolinecolor":"rgba(0, 0, 0, 1.000)","anchor":"x1"},"yaxis1":{"showticklabels":true,"gridwidth":0.5,"tickvals":[1.0,1.5,2.0,2.5,3.0],"visible":true,"ticks":"inside","range":[1.0,3.0],"domain":[0.5376202974628171,0.9901574803149606],"tickmode":"array","linecolor":"rgba(0, 0, 0, 1.000)","showgrid":true,"title":"","mirror":false,"tickangle":0,"showline":true,"gridcolor":"rgba(0, 0, 0, 0.100)","titlefont":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":15},"tickcolor":"rgb(0, 0, 0)","ticktext":["1.0","1.5","2.0","2.5","3.0"],"zeroline":false,"type":"-","tickfont":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":11},"zerolinecolor":"rgba(0, 0, 0, 1.000)","anchor":"x1"},"annotations":[],"plot_bgcolor":"rgba(255, 255, 255, 1.000)","showlegend":true,"xaxis1":{"showticklabels":true,"gridwidth":0.5,"tickvals":[1.0,1.5,2.0,2.5,3.0],"visible":true,"ticks":"inside","range":[1.0,3.0],"domain":[0.05100612423447069,0.9934383202099737],"tickmode":"array","linecolor":"rgba(0, 0, 0, 1.000)","showgrid":true,"title":"","mirror":false,"tickangle":0,"showline":true,"gridcolor":"rgba(0, 0, 0, 0.100)","titlefont":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":15},"tickcolor":"rgb(0, 0, 0)","ticktext":["1.0","1.5","2.0","2.5","3.0"],"zeroline":false,"type":"-","tickfont":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":11},"zerolinecolor":"rgba(0, 0, 0, 1.000)","anchor":"y2"},"legend":{"bordercolor":"rgba(0, 0, 0, 1.000)","bgcolor":"rgba(255, 255, 255, 1.000)","font":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":11},"y":1.0,"x":1.0},"margin":{"l":0,"b":20,"r":0,"t":20},"width":600});
    </script>

    </body>
</html>
```
but not with `plotlyjs()`:
```html
<html>
<head>
     <script src="C:\Users\Daniel\.julia\v0.6\PlotlyJS\assets\plotly-latest.min.js"></script>
</head>
<body>
     <div id="fe032bd9-f90f-4b23-9e6d-6c5c21c77542" class="plotly-graph-div"></div>

<script>
    window.PLOTLYENV=window.PLOTLYENV || {};
    window.PLOTLYENV.BASE_URL="https://plot.ly";
    Plotly.newPlot('fe032bd9-f90f-4b23-9e6d-6c5c21c77542', [{"showlegend":true,"mode":"lines","xaxis":"x1","colorbar":{"title":""},"line":{"color":"rgba(0, 154, 250, 1.000)","width":1,"dash":"solid","shape":"linear"},"y":[1,2,3],"type":"scatter","name":"y1","yaxis":"y1","x":[1,2,3]},{"showlegend":true,"mode":"lines","xaxis":"x2","colorbar":{"title":""},"line":{"color":"rgba(0, 154, 250, 1.000)","width":1,"dash":"solid","shape":"linear"},"y":[3,4,5],"type":"scatter","name":"y1","yaxis":"y2","x":[1,2,3]}],
               {"paper_bgcolor":"rgba(255, 255, 255, 1.000)","height":400,"yaxis2":{"showticklabels":true,"gridwidth":0.5,"tickvals":[3.0,3.5,4.0,4.5,5.0],"visible":true,"ticks":"inside","range":[3.0,5.0],"domain":[0.03762029746281716,0.4901574803149606],"tickmode":"array","linecolor":"rgba(0, 0, 0, 1.000)","showgrid":true,"title":"","mirror":false,"tickangle":0,"showline":true,"gridcolor":"rgba(0, 0, 0, 0.100)","titlefont":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":15},"tickcolor":"rgb(0, 0, 0)","ticktext":["3.0","3.5","4.0","4.5","5.0"],"zeroline":false,"type":"-","tickfont":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":11},"zerolinecolor":"rgba(0, 0, 0, 1.000)","anchor":"x1"},"yaxis1":{"showticklabels":true,"gridwidth":0.5,"tickvals":[1.0,1.5,2.0,2.5,3.0],"visible":true,"ticks":"inside","range":[1.0,3.0],"domain":[0.5376202974628171,0.9901574803149606],"tickmode":"array","linecolor":"rgba(0, 0, 0, 1.000)","showgrid":true,"title":"","mirror":false,"tickangle":0,"showline":true,"gridcolor":"rgba(0, 0, 0, 0.100)","titlefont":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":15},"tickcolor":"rgb(0, 0, 0)","ticktext":["1.0","1.5","2.0","2.5","3.0"],"zeroline":false,"type":"-","tickfont":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":11},"zerolinecolor":"rgba(0, 0, 0, 1.000)","anchor":"x1"},"annotations":[],"plot_bgcolor":"rgba(255, 255, 255, 1.000)","showlegend":true,"xaxis1":{"showticklabels":true,"gridwidth":0.5,"tickvals":[1.0,1.5,2.0,2.5,3.0],"visible":true,"ticks":"inside","range":[1.0,3.0],"domain":[0.05100612423447069,0.9934383202099737],"tickmode":"array","linecolor":"rgba(0, 0, 0, 1.000)","showgrid":true,"title":"","mirror":false,"tickangle":0,"showline":true,"gridcolor":"rgba(0, 0, 0, 0.100)","titlefont":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":15},"tickcolor":"rgb(0, 0, 0)","ticktext":["1.0","1.5","2.0","2.5","3.0"],"zeroline":false,"type":"-","tickfont":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":11},"zerolinecolor":"rgba(0, 0, 0, 1.000)","anchor":"y2"},"legend":{"bordercolor":"rgba(0, 0, 0, 1.000)","bgcolor":"rgba(255, 255, 255, 1.000)","font":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":11},"y":1.0,"x":1.0},"margin":{"l":0,"b":20,"r":0,"t":20},"width":600}, {showLink: false});

 </script>

</body>
</html>
```

**Comment by daschw** [comment]
For me too!

**Comment by cstjean** [comment]
Works great, even with DateTime axes. Thank you!

**Pull request merged**